### PR TITLE
Add sshfp support

### LINKED ISSUE #489: Add SSHFP DNS support
This was requested in #476, seems pretty doable.

### PR DESCRIPTION
**Pull request by claudijd**
Fixes #489, which was requested in #476

### CODE REVIEW

**Comment by bhourigan** [comment]
Great feature!

**Comment by claudijd** [comment]
For a preview, it will simply add the following block to the output (JSON for instance)...

    "ssh_fps": [
      {
        "fptype": "sha1",
        "algo": "ed25519",
        "hex": "69:ac:08:0c:cf:6c:d5:2f:47:88:37:3b:d4:dc:a2:17:31:e6:97:13"
      },
      {
        "fptype": "sha1",
        "algo": "ecdsa",
        "hex": "7c:4b:9b:91:05:d6:a0:d7:aa:cf:44:53:4a:78:00:fc:10:46:66:83"
      },
      {
        "fptype": "sha256",
        "algo": "ed25519",
        "hex": "7c:ae:4f:f9:42:89:9f:8e:15:5b:fc:67:5e:72:e4:14:6a:1b:f4:10:79:77:fe:73:c6:cf:fa:8f:3f:da:8f:c3"
      },
      {
        "fptype": "sha256",
        "algo": "ecdsa",
        "hex": "cb:64:93:b1:0e:11:03:ff:1d:ba:b8:69:89:cf:a9:6f:a5:23:70:ac:33:ef:e6:d4:68:a5:f7:0b:8d:32:38:69"
      }
    ],

It could probably use some better standarization and feedback on what exact keys we should use and how we want to organize things.  I think the big value here is being able to look at the SSHFP records and sync them up with the keys that were offered.  If they are say a mismatch, we can tell that a client expecting SSHFP verification to fail and could make that an item for people to fix up.

**Comment by claudijd** [comment]
Here's an example of why we switched to hex for SHA256 in this, it allows us to by eye, compare key fingerprints with sshfp fingerprints we get via DNS....

    "keys": {
      "dsa": {
        "raw": "ssh-dss AAAAB3NzaC1kc3MAAACBAIpTNR55FE+QBSjp9JtMWuJqG+wtx35fpE3s2WHd/pPqfkVCcBP+U4ytps+/8IE0QUnfadgrCGl8N33l/tuUmTr7MKOwOLX/+70LsXOkKLXphO1L8NwTjr724Zguymy4IoLBH8onedePBuj/YWEYz9j0n/shwahRnA7nYensrj3LAAAAFQCIm2MH0if9fH6HU9ggnbURfSIwyQAAAIA3D0Pyq/6Oe2TuBkKNJdcCta9ISxJQlk+kk4+hjc0QBFNzUjHBZfyncJR+3i2dfjRVF7ri8WA3N2MYviUEJovnxIqYSkX0rvRDPGzAf0yodX8SZOv+pj3PKg62VoESwYN+vocWW1snYDk3ix/GLxNg+Q8Zgf7hAPOYigbBLMMQJwAAAIBVyy4V5lidbFeeL3FonZo64b9Ddw8lTabHKqh9r6q0HLKHabRU1yFMSRu9NyA8IK44/UuvbeJOmJNtXHvBm43KjhjeLuEuRDp9IrYbS1JSuNFREANO/ovaS4AV+O1QSeuBKBc2OTw6KvGKJuSgWaZeArl1WxpHobzVmOByq5tGsQ==",
        "length": 1024,
        "fingerprints": {
          "md5": "26:05:8c:e6:a0:34:11:b1:f8:de:dc:3c:72:2a:6b:14",
          "sha1": "49:3a:9e:6a:4b:50:78:b1:d0:c5:42:4a:ec:f8:17:ea:54:e1:df:df",
          "sha256": "06:4b:9d:d1:08:05:06:9e:b5:08:bd:08:7a:37:db:61:fd:a2:10:71:38:92:41:12:de:d3:cc:db:aa:fd:6c:b3"
        }
      },
      "rsa": {
        "raw": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDDGtFi3OFEkHY2b0F0gLMtHvPQMvOMH+0JFJzaP8qw6UUrpmaGTzSfaqgA5TtM46mP+VIpipIspHcltzAE43IFZTbD88EWz/3v8SwvD3PBX7JQSN4Hgs+RuQyFJnPrYy/Ms1IPxHXFIkTtrZYZSg/jJ1T4VdSNxhzLhDMqPmTSkef2WQPlEDd0q+SHnlCwiweHQthmLvnTA18OhZxFI+vzdfjYB2+DYmyQuEHaQVA8Y5rCEi9wf5dc1ZiojQA7zpk5bpGaajncLLPuQrvsxl6DPs8nBtbIxhlRMaGeCCSZCWJ7xIymJkNLyYjg3Up6LOZXc1eeSdLlxhiNWuho7wD/",
        "length": 2048,
        "fingerprints": {
          "md5": "82:a3:df:33:18:4d:96:53:3e:f1:43:8d:8a:4d:9b:d7",
          "sha1": "db:74:48:17:e8:d6:ac:20:27:e6:62:9a:ac:7f:0f:c1:75:0f:65:88",
          "sha256": "a6:1d:b0:2b:9b:26:ca:48:66:3c:32:72:82:1b:45:17:73:c7:cd:1e:9a:41:2f:5a:09:99:4e:c8:f8:73:8c:79"
        }
      },
      "ecdsa-sha2-nistp256": {
        "raw": "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBJt5zFtcGMQJbWMu9re0aNYjOlcGaaKmDhWs/e5ZfJ9rBm+Y+7BwxOWga92fVFuiLWmcqnBxM7kkNNKV/QdCTIA=",
        "length": 520,
        "fingerprints": {
          "md5": "cc:e3:6c:6f:6d:a0:2a:3a:d2:5d:15:b1:93:e1:f7:90",
          "sha1": "7c:4b:9b:91:05:d6:a0:d7:aa:cf:44:53:4a:78:00:fc:10:46:66:83",
          "sha256": "cb:64:93:b1:0e:11:03:ff:1d:ba:b8:69:89:cf:a9:6f:a5:23:70:ac:33:ef:e6:d4:68:a5:f7:0b:8d:32:38:69"
        }
      },
      "ed25519": {
        "raw": "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIPtol6MWE48rbDz/UXsn9sjuQd2mDKhiebqcqAae3BZm",
        "length": 256,
        "fingerprints": {
          "md5": "f1:ce:b2:b9:da:1f:36:3c:2e:96:4a:fa:95:f5:b5:97",
          "sha1": "69:ac:08:0c:cf:6c:d5:2f:47:88:37:3b:d4:dc:a2:17:31:e6:97:13",
          "sha256": "7c:ae:4f:f9:42:89:9f:8e:15:5b:fc:67:5e:72:e4:14:6a:1b:f4:10:79:77:fe:73:c6:cf:fa:8f:3f:da:8f:c3"
        }
      }
    },
    "ssh_fps": [
      {
        "fptype": "sha1",
        "algo": "ed25519",
        "hex": "69:ac:08:0c:cf:6c:d5:2f:47:88:37:3b:d4:dc:a2:17:31:e6:97:13"
      },
      {
        "fptype": "sha1",
        "algo": "ecdsa",
        "hex": "7c:4b:9b:91:05:d6:a0:d7:aa:cf:44:53:4a:78:00:fc:10:46:66:83"
      },
      {
        "fptype": "sha256",
        "algo": "ed25519",
        "hex": "7c:ae:4f:f9:42:89:9f:8e:15:5b:fc:67:5e:72:e4:14:6a:1b:f4:10:79:77:fe:73:c6:cf:fa:8f:3f:da:8f:c3"
      },
      {
        "fptype": "sha256",
        "algo": "ecdsa",
        "hex": "cb:64:93:b1:0e:11:03:ff:1d:ba:b8:69:89:cf:a9:6f:a5:23:70:ac:33:ef:e6:d4:68:a5:f7:0b:8d:32:38:69"
      }
    ],

File: spec/ssh_scan/ssh_fp_spec.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+          "hex"=>
+          "cb:64:93:b1:0e:11:03:ff:1d:ba:b8:69:89:cf:a9:6f:a5:23:70:ac:33:ef:e6:d4:68:a5:f7:0b:8d:32:38:69"},
+         {"algo"=>"ed25519",
+          "fptype"=>"sha1",
+          "hex"=>"69:ac:08:0c:cf:6c:d5:2f:47:88:37:3b:d4:dc:a2:17:31:e6:97:13"},
+         {"algo"=>"ed25519",
+          "fptype"=>"sha256",
+          "hex"=>
+           "7c:ae:4f:f9:42:89:9f:8e:15:5b:fc:67:5e:72:e4:14:6a:1b:f4:10:79:77:fe:73:c6:cf:fa:8f:3f:da:8f:c3"}
+        ].sort_by { |k| k["hex"] }
```
Comment (claudijd) [inline spec/ssh_scan/ssh_fp_spec.rb]: sorting here is just to avoid ordering bugs in the spec runs

File: lib/ssh_scan/public_key.rb [hunk 2]
```diff
@@ -39,10 +39,11 @@ def fingerprint_md5
 
       def fingerprint_sha1
         SSHKey.sha1_fingerprint(@key_string)
-      end    
+      end
 
       def fingerprint_sha256
-        SSHKey.sha256_fingerprint(@key_string)
+        # We're translating this to hex because the SSHKEY default isn't as useful for comparing with SSHFP records
+        Base64.decode64(SSHKey.sha256_fingerprint(@key_string)).hexify(:delim => ":")
```
Comment (claudijd) [inline lib/ssh_scan/public_key.rb]: I can see why sshkey base64'd this before, it's simply a shorter format to express the same data.  I think hex is just easier to see if you're manually reviewing.  We can then write rules to catch weird violations.

**Comment by claudijd** [comment]
I changed the name of sshfps to dns_keys, it just seemed more forthcoming about where the keys were coming from if not the direct SSH connection we were probing...

    "keys": {
      "dsa": {
        "raw": "ssh-dss AAAAB3NzaC1kc3MAAACBAIpTNR55FE+QBSjp9JtMWuJqG+wtx35fpE3s2WHd/pPqfkVCcBP+U4ytps+/8IE0QUnfadgrCGl8N33l/tuUmTr7MKOwOLX/+70LsXOkKLXphO1L8NwTjr724Zguymy4IoLBH8onedePBuj/YWEYz9j0n/shwahRnA7nYensrj3LAAAAFQCIm2MH0if9fH6HU9ggnbURfSIwyQAAAIA3D0Pyq/6Oe2TuBkKNJdcCta9ISxJQlk+kk4+hjc0QBFNzUjHBZfyncJR+3i2dfjRVF7ri8WA3N2MYviUEJovnxIqYSkX0rvRDPGzAf0yodX8SZOv+pj3PKg62VoESwYN+vocWW1snYDk3ix/GLxNg+Q8Zgf7hAPOYigbBLMMQJwAAAIBVyy4V5lidbFeeL3FonZo64b9Ddw8lTabHKqh9r6q0HLKHabRU1yFMSRu9NyA8IK44/UuvbeJOmJNtXHvBm43KjhjeLuEuRDp9IrYbS1JSuNFREANO/ovaS4AV+O1QSeuBKBc2OTw6KvGKJuSgWaZeArl1WxpHobzVmOByq5tGsQ==",
        "length": 1024,
        "fingerprints": {
          "md5": "26:05:8c:e6:a0:34:11:b1:f8:de:dc:3c:72:2a:6b:14",
          "sha1": "49:3a:9e:6a:4b:50:78:b1:d0:c5:42:4a:ec:f8:17:ea:54:e1:df:df",
          "sha256": "06:4b:9d:d1:08:05:06:9e:b5:08:bd:08:7a:37:db:61:fd:a2:10:71:38:92:41:12:de:d3:cc:db:aa:fd:6c:b3"
        }
      },
      "ecdsa-sha2-nistp256": {
        "raw": "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBJt5zFtcGMQJbWMu9re0aNYjOlcGaaKmDhWs/e5ZfJ9rBm+Y+7BwxOWga92fVFuiLWmcqnBxM7kkNNKV/QdCTIA=",
        "length": 520,
        "fingerprints": {
          "md5": "cc:e3:6c:6f:6d:a0:2a:3a:d2:5d:15:b1:93:e1:f7:90",
          "sha1": "7c:4b:9b:91:05:d6:a0:d7:aa:cf:44:53:4a:78:00:fc:10:46:66:83",
          "sha256": "cb:64:93:b1:0e:11:03:ff:1d:ba:b8:69:89:cf:a9:6f:a5:23:70:ac:33:ef:e6:d4:68:a5:f7:0b:8d:32:38:69"
        }
      },
      "ed25519": {
        "raw": "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIPtol6MWE48rbDz/UXsn9sjuQd2mDKhiebqcqAae3BZm",
        "length": 256,
        "fingerprints": {
          "md5": "f1:ce:b2:b9:da:1f:36:3c:2e:96:4a:fa:95:f5:b5:97",
          "sha1": "69:ac:08:0c:cf:6c:d5:2f:47:88:37:3b:d4:dc:a2:17:31:e6:97:13",
          "sha256": "7c:ae:4f:f9:42:89:9f:8e:15:5b:fc:67:5e:72:e4:14:6a:1b:f4:10:79:77:fe:73:c6:cf:fa:8f:3f:da:8f:c3"
        }
      },
      "rsa": {
        "raw": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDDGtFi3OFEkHY2b0F0gLMtHvPQMvOMH+0JFJzaP8qw6UUrpmaGTzSfaqgA5TtM46mP+VIpipIspHcltzAE43IFZTbD88EWz/3v8SwvD3PBX7JQSN4Hgs+RuQyFJnPrYy/Ms1IPxHXFIkTtrZYZSg/jJ1T4VdSNxhzLhDMqPmTSkef2WQPlEDd0q+SHnlCwiweHQthmLvnTA18OhZxFI+vzdfjYB2+DYmyQuEHaQVA8Y5rCEi9wf5dc1ZiojQA7zpk5bpGaajncLLPuQrvsxl6DPs8nBtbIxhlRMaGeCCSZCWJ7xIymJkNLyYjg3Up6LOZXc1eeSdLlxhiNWuho7wD/",
        "length": 2048,
        "fingerprints": {
          "md5": "82:a3:df:33:18:4d:96:53:3e:f1:43:8d:8a:4d:9b:d7",
          "sha1": "db:74:48:17:e8:d6:ac:20:27:e6:62:9a:ac:7f:0f:c1:75:0f:65:88",
          "sha256": "a6:1d:b0:2b:9b:26:ca:48:66:3c:32:72:82:1b:45:17:73:c7:cd:1e:9a:41:2f:5a:09:99:4e:c8:f8:73:8c:79"
        }
      }
    },
    "dns_keys": [
      {
        "fptype": "sha1",
        "algo": "ed25519",
        "hex": "69:ac:08:0c:cf:6c:d5:2f:47:88:37:3b:d4:dc:a2:17:31:e6:97:13"
      },
      {
        "fptype": "sha1",
        "algo": "ecdsa",
        "hex": "7c:4b:9b:91:05:d6:a0:d7:aa:cf:44:53:4a:78:00:fc:10:46:66:83"
      },
      {
        "fptype": "sha256",
        "algo": "ed25519",
        "hex": "7c:ae:4f:f9:42:89:9f:8e:15:5b:fc:67:5e:72:e4:14:6a:1b:f4:10:79:77:fe:73:c6:cf:fa:8f:3f:da:8f:c3"
      },
      {
        "fptype": "sha256",
        "algo": "ecdsa",
        "hex": "cb:64:93:b1:0e:11:03:ff:1d:ba:b8:69:89:cf:a9:6f:a5:23:70:ac:33:ef:e6:d4:68:a5:f7:0b:8d:32:38:69"
      }
    ],

**Comment by exploide** [comment]
Looks great. No objections so far.

**Pull request merged**